pkg/sentry/devices/nvproxy/frontend.go

```diff
@@ -587,6 +587,31 @@ func ctrlClientSystemGetBuildVersion(fi *frontendIoctlState, ioctlParams *nvgpu.
 	return n, nil
 }
 
+func ctrlDevGpuGetClasslist(fi *frontendIoctlState, ioctlParams *nvgpu.NVOS54Parameters) (uintptr, error) {
+	var ctrlParams nvgpu.NV0080_CTRL_CMD_GPU_GET_CLASSLIST_PARAMS
+	if ctrlParams.SizeBytes() != int(ioctlParams.ParamsSize) {
+		return 0, linuxerr.EINVAL
+	}
+
+	if _, err := ctrlParams.CopyIn(fi.t, addrFromP64(ioctlParams.Params)); err != nil {
+		return 0, err
+	}
+
+	// This command has two modes. If the classList pointer is NULL, only simple command handling
+	// is required; see src/common/sdk/nvidia/inc/ctrl/ctrl0080gpu.h.
+	if ctrlParams.ClassList == 0 {
+		return rmControlSimple(fi, ioctlParams);
+	}
+
+	// classList pointer is not NULL. Do classList buffer management.
+	classList := make([]uint32, ctrlParams.NumClasses)
```

Comment:
@manninglucas bought it up during internal code review that we may want to sanitize this user-defined variable to avoid huge malicious allocations.

Reply:
IIUC, the driver imposes a limit of RMAPI_PARAM_COPY_MAX_PARAMS_SIZE = 1 MB (src/nvidia/inc/kernel/rmapi/param_copy.h) per copied "parameter", in src/nvidia/src/kernel/rmapi/param_copy.c:rmapiParamsAcquire(); we can impose the same limit here.

Reply:
Makes sense, will add to PR 👍

Reply:
Used the existing max classlist constant from `ctrl0080gpu.h`. Even though it's a constant defined for the v2 command, it seems appropriate for this use case too, as the set of classes is the same.

Reply:
Actually `NV0080_CTRL_GPU_CLASSLIST_MAX_SIZE` is not the right thing to use here. Is it the list size used by deviceCtrlCmdGpuGetClasslistV2_IMPL(). And it is used to impose a limit for some sort of a generated list in src/nvidia/generated/g_gpu_class_list.c:gpuGetClassDescriptorList_GA104().

I think as Jamie pointed, we should use the 1 MB limit imposed by src/nvidia/src/kernel/rmapi/param_copy.c:rmapiParamsAcquire() via `RMAPI_PARAM_COPY_MAX_PARAMS_SIZE` because that is what the driver does. Currently the limit is too low. I am working on a similar change, will fix this too.